### PR TITLE
python3Packages.attrs: 20.3.0 -> 21.2.0

### PR DESCRIPTION
**Pull request by lukegb**
###### Motivation for this change

https://pypi.org/project/attrs

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by dotlambda — COMMENTED** [review]
Can you add `pythonImportsCheck` and maybe switch to `pytestCheckHook`?

**Comment by lukegb** [comment]
> Can you add `pythonImportsCheck` and maybe switch to `pytestCheckHook`?

Yes. The second half is actually... a little bit more convoluted, but I've tried to work around it by using a separate derivation which can depend on both attrs and pytest (since attrs is a dep of pytest, and I can't break the pytest->attrs dep very easily)

cc @jonringer @FRidh because this is a little unconventional in terms of Python packaging and I would appreciate some feedback. I think this is a better state than just leaving it at `doCheck = false;`, but I'm unsure if there's a better way.

@roberth suggested on IRC maybe overriding pytest's dep on attrs to set doCheck = false;, but I _believe_ that's a bad idea because then there's a strong possibility of attrs ending up on the Python path multiple times - even though it's "harmless" in this case I don't really want to reason about that.

**Comment by lukegb** [comment]
@ofborg build python3.pkgs.attrs python3.pkgs.attrs.tests

**Comment by jonringer** [comment]
> cc @jonringer @FRidh because this is a little unconventional in terms of Python packaging and I would appreciate some feedback. I think this is a better state than just leaving it at `doCheck = false;`, but I'm unsure if there's a better way.

I think this is a clever and better way of achieving the goal of testing

> @roberth suggested on IRC maybe overriding pytest's dep on attrs to set doCheck = false;, but I believe that's a bad idea because then there's a strong possibility of attrs ending up on the Python path multiple times - even though it's "harmless" in this case I don't really want to reason about that.

We do this for `hypothesis`, but I think it's just because another paradigm wasn't common.

File: pkgs/development/python-modules/attrs/default.nix [hunk 1]
```diff
@@ -1,27 +1,36 @@
...
+    # Install tests as the tests output.
+    mkdir $testout
+    cp -R tests $testout/tests
   '';
 
-  # To prevent infinite recursion with pytest
+  pythonImportsCheck = [ "attr" ];
+
+  # pytest depends on attrs, so we can't do this out-of-the-box.
+  # Instead, we do this as a passthru.tests test instead.
```
Comment (mweinelt) [inline pkgs/development/python-modules/attrs/default.nix:27]:
```suggestion
  # Instead, we do this as a passthru.tests test.
```

Comment (mweinelt) [reply pkgs/development/python-modules/attrs/default.nix:27]: Fixed locally.

**Comment by mweinelt** [comment]
Merged into my local working tree for python-unstable. Thank you.

**Pull request closed without merging**